allennlp/modules/seq2seq_encoders/pytorch_seq2seq_wrapper.py

```diff
@@ -29,76 +29,93 @@ class PytorchSeq2SeqWrapper(Seq2SeqEncoder):
 
     def forward(self,  # pylint: disable=arguments-differ
                 inputs: torch.Tensor,
                 mask: torch.Tensor,
                 hidden_state: torch.Tensor = None) -> torch.Tensor:
 
+        if self._stateful and mask is None:
+            raise ValueError("Always pass a mask with stateful RNNs.")
+        if self._stateful and hidden_state is not None:
+            raise ValueError("Stateful RNNs provide their own initial hidden_state.")
+
         if mask is None:
             return self._module(inputs, hidden_state)[0]
 
-        # In some circumstances you may have sequences of zero length.
-        # ``pack_padded_sequence`` requires all sequence lengths to be > 0, so here we
-        # adjust the ``mask`` so that every sequence has length at least 1. Then after
-        # running the RNN we zero out the corresponding rows in the result.
-
-        # First count how many sequences are empty.
         batch_size, total_sequence_length = mask.size()
-        num_valid = torch.sum(mask[:, 0]).int().data[0]
 
-        # Force every sequence to be length at least one. Need to `.clone()` the mask
-        # to avoid a RuntimeError from shared storage.
-        if num_valid < batch_size:
-            mask = mask.clone()
-            mask[:, 0] = 1
-
-        sequence_lengths = get_lengths_from_binary_sequence_mask(mask)
-        sorted_inputs, sorted_sequence_lengths, restoration_indices = sort_batch_by_length(inputs,
-                                                                                           sequence_lengths)
-        packed_sequence_input = pack_padded_sequence(sorted_inputs,
-                                                     sorted_sequence_lengths.data.tolist(),
-                                                     batch_first=True)
-
-        # Actually call the module on the sorted PackedSequence.
-        packed_sequence_output, _ = self._module(packed_sequence_input, hidden_state)
+        packed_sequence_output, final_states, restoration_indices, num_valid = \
+            self.sort_and_run_forward(self._module, inputs, mask, hidden_state)
+
         unpacked_sequence_tensor, _ = pad_packed_sequence(packed_sequence_output, batch_first=True)
 
-        # We sorted by length, so if there are invalid rows that need to be zeroed out
-        # they will be at the end.
+        # Some RNNs (GRUs) only return one state as a Tensor.  Others (LSTMs) return two.
+        # If one state, use a single element list to handle in a consistent manner below.
+        if not isinstance(final_states, (list, tuple)) and self._stateful:
+            final_states = [final_states]
+
+        # Add back invalid rows.
         if num_valid < batch_size:
-            unpacked_sequence_tensor[num_valid:, :, :] = 0.
+            _, length, output_dim = unpacked_sequence_tensor.size()
+            zeros = unpacked_sequence_tensor.data.new(batch_size - num_valid, length, output_dim).fill_(0)
```

Comment:
You do this operation enough that I'd probably just add a method to `util` for it.  Something like `util.add_rows(tensor, dim, num_rows, fill_value)`.